pubs/bibstruct.py

```diff
@@ -56,26 +65,51 @@ def valid_citekey(citekey):
     # FIXME: a bit crude, but efficient for now (and allows unicode citekeys)
     return not '/' in citekey
 
-
-def generate_citekey(bibdata):
+class CitekeyFormatter(Formatter):
+    def __init__(self):
+        super(CitekeyFormatter, self).__init__()
+
+    def format_field(self, val, fmt):
+        if len(fmt) > 0 and fmt[0] == 'u':
+            s = str(val).upper()
+            fmt = fmt[1:]
+        elif len(fmt) > 0 and fmt[0] == 'l':
+            s = str(val).lower()
+            fmt = fmt[1:]
+        else:
+            s = val
+        return str2citekey(s.__format__(fmt))
+
+
+    def get_value(self, key, args, entry):
+        if isinstance(key, (str, unicode)):
+            okey = key
+            if key == 'author' and not 'author' in entry:
+                key = 'editor'
+            elif key == 'editor' and not 'editor' in entry:
+                key = 'author'
+
+            if key == 'first_word' and 'title' in entry:
+                return entry['title'].split(' ')[0]
```

Comment:
A bit brittle. Won't work properly on title with colons, commas (e.g. "Long, Windy Title", "Robots: the Next Revolution").